MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java

```diff
@@ -0,0 +1,243 @@
+    }
+
+    public void onSurfaceCreated() {
+        Log.d(TAG, "Initializing state");
+
+        GLES20.glClearColor(1.0f, 0.0f, 0.0f, 0.0f);
+
+        int vertexShader = compileShader(GLES20.GL_VERTEX_SHADER, VERTEX_SHADER);
+        int fragmentShader = compileShader(GLES20.GL_FRAGMENT_SHADER, FRAGMENT_SHADER);
+        program = linkProgram(vertexShader, fragmentShader);
+
+        // Get vertex and texture handles
+        aPositionHandle = GLES20.glGetAttribLocation(program, "a_Position");
+        aTextureHandle = GLES20.glGetAttribLocation(program, "a_TextureCoordinates");
+        uMatrixHandle = GLES20.glGetUniformLocation(program, "u_Matrix");
+        uTextureUnitHandle = GLES20.glGetUniformLocation(program, "u_TextureUnit");
+
+        // Validate the program
+        GLES20.glValidateProgram(program);
+        final int[] validateStatus = new int[1];
+        GLES20.glGetProgramiv(program, GLES20.GL_VALIDATE_STATUS, validateStatus, /* offset = */ 0);
+        Log.v(TAG, "Program validation results: " + validateStatus[0]
+                + "\nLog: " + GLES20.glGetProgramInfoLog(program));
+
+        GLES20.glUseProgram(program);
+
+        // Create and bind textures
+        int[] textureIds = new int[1];
+        GLES20.glGenTextures(/* n = */ 1, textureIds, 0);
+        textureID = textureIds[0];
+        GLES20.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, textureID);
+
+        // Set texture parameters
+        GLES20.glTexParameterf(
+                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
+                GLES20.GL_TEXTURE_MIN_FILTER,
+                GLES20.GL_NEAREST
+        );
+        GLES20.glTexParameterf(
+                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
+                GLES20.GL_TEXTURE_MAG_FILTER,
+                GLES20.GL_LINEAR
+        );
+        GLES20.glTexParameteri(
+                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
+                GLES20.GL_TEXTURE_WRAP_S,
+                GLES20.GL_CLAMP_TO_EDGE
+        );
+        GLES20.glTexParameteri(
+                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
+                GLES20.GL_TEXTURE_WRAP_T,
+                GLES20.GL_CLAMP_TO_EDGE
+        );
+    }
+
+    private static int linkProgram(int vertexShader, int fragmentShader) {
+        Log.d(TAG, "Linking program");
+
+        // Create the program
+        final int program = GLES20.glCreateProgram();
+        if (program == 0) {
+            Log.w(TAG, "Could not create new program");
+            return 0;
+        }
+
+        // Attach shaders to program and link them together
+        GLES20.glAttachShader(program, vertexShader);
+        GLES20.glAttachShader(program, fragmentShader);
+        GLES20.glLinkProgram(program);
+
+        // Verify linking status
+        final int[] linkStatus = new int[1];
+        GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, linkStatus, /* offset = */ 0);
+        if (linkStatus[0] == 0) {
+            // Delete the program if linking failed
```

Comment:
Throw a RuntimeException in that case with the info log - that will help you debug your shader. Same for the shader compilation status below.